src/back/mod.rs

```diff
@@ -13,6 +18,39 @@ pub mod spv;
 #[cfg(feature = "wgsl-out")]
 pub mod wgsl;
 
+/// How should code generated by Naga do indexing bounds checks?
+///
+/// When a vector, matrix, or array index is out of bounds—either negative, or
+/// greater than or equal to the number of elements in the type—WGSL requires
+/// that some other index of the implementation's choice that is in bounds is
+/// used instead. (There are no types with zero elements.)
+///
+/// Although the WebGPU platform must behave as WGSL indicates, non-web users
+/// may have other needs. Some may want to avoid the overhead of a bounds check;
+/// others may want the checks, but prefer other behaviors.
+#[derive(Clone, Copy, Debug)]
+#[cfg_attr(feature = "serialize", derive(Serialize))]
+#[cfg_attr(feature = "deserialize", derive(Deserialize))]
+pub enum IndexBoundsCheckPolicy {
+    /// Replace out-of-bounds indexes with some arbitrary in-bounds index.
+    /// (This does not necessarily mean clamping.)
+    Restrict,
+
+    /// When presented with an out-of-bounds index check, reads return zero, and
+    /// writes have no effect.
+    ReadZeroSkipWrite,
+
+    /// Out-of-bounds indexes are undefined behavior. Generate the fastest code
+    /// possible.
+    UndefinedBehavior,
+}
+
+impl Default for IndexBoundsCheckPolicy {
```

Comment:
This is a tough one to default... and I guess it has to be defaulted, since it's a part of `Options`, where a default is expected.
I think, instead of following the WebGPU requirements here, the default should follow the basic idea of explicit opt-in into more work or features. Since restricting or zeroing is extra code, it's a feature. So the default should be `UndefinedBehavior`.

Reply:
I'd argue that undefined behavior is something people should have to opt into. It seems to me that the "damn the torpedos, full speed ahead" approach of `UndefinedBehavior` is the feature, and that predictable behavior, following the spec, is the boring status quo, that you get if you're not really paying attention or a newbie.

Reply:
While I do want to push a bit for a safe default here, it's your call. If you think UB by default is a necessity, then I'm happy to make the change.

Reply:
Yes, I'm convinced Naga should be principled in opting into extra work, here, not safey.
To add more justification, Naga by itself can't generally provide safety, anyway. It relies on the host API, for example to pass the buffer sizes into internal data (like we do in Metal for the `ArrayLength`). The only layer that *can* guarantee safety is `wgpu`, and it's defaulting to safe behaviors. This defaulting includes configuring Naga in a way it needs.

Reply:
Okay, sounds good.